Cesmdev update

mvertens opened the pull request:
Updates from cesmdev master

1) Add in option for CLMGSWP3 for datm. Also add in capability for driver
namelist generation to merge CAM and CLM drv_flds_in namelists
together, and abort on a conflict. Add in testing of driver
namelist. Work on SSP test and add SPL testcase. Add option to
Namelist.pm->merge_nl to die on conflict. change notes for Namelist.pm
to be "!#" instead of "#!". Fix namelist test to work. Remove messing
with drv_flds_in file in preview_namelists.

2) Changes to deposition tables in seq_drydep_mod for additional VOC type
species.

3) Remove inputdata files from usermods that referenced glade so
that tests will pass on tests outside NCAR.

4) Rework the vertical gradient calculator for lnd to glc remapping in a
few ways:

(a) Adds a limiter to the vertical gradient calculator
(vertical_gradient_calculator_2nd_order), restricting gradients such
that the values at the interface between two elevation classes must
lie in between the mean values in those two elevation classes. Bill
Lipscomb provided the initial implementation for this change.

(b) Significantly reworks the API and internals of the vertical gradient
calculator, as well as related unit tests.

(c) Adds a factory module to create vertical gradient calculators -
making the vertical gradient calculator's constructor independent of
attribute vectors

(c) Adds a script for plotting vertical gradients

Changes (b), (c) and (d) were not necessary to accomplish (a1). Rather,
these changes and additions were made in order to support the
implementation of a new, more complex vertical gradient calculator
(which ensured continuity at elevation class interfaces via a matrix
solve for the gradients). The results of this new vertical gradient
calculator were not as good as we had hoped, so Bill Sacks backed that out -
but kept in place the changes in (b), (c) and (d) because these seemed
to be an improvement to the design.

Changes answers for all compsets with an active GLC (i.e., with CISM). 
This only feeds back on the climate if two-way evolution is on
otherwise, this is a diagnostic-only change. 

5) Also fixed problems with the short term archiver, etc that were needed
to address clm45_aux testing issues

Test suite: scripts_regression_tests
Test baseline: NA
Test namelist changes: NA
Test status: all passed

Fixes: [CIME Github issue #]

User interface changes?: None

Code review:

jgfouca commented on cime_config/cesm/allactive/testlist_allactive.xml:
```diff
@@ -400,4 +384,256 @@
       </options>
     </machines>
   </test>
+=======
```
These ==== shoudn't be here.

mvertens replied:
Agreed. This is fixed.

jgfouca commented on driver_cpl/bld/build-namelist:
```diff
@@ -176,7 +176,7 @@ EOF
 
 # The Build::NamelistDefinition module provides utilities to validate that the output
-# namelists are consistent with the namelist definition file
+# namelists are consistent with the namelist definition files
```
Isn't the perl version of build-namelist not used anymore?

mvertens replied:
Yes - the perl version of build-namelist is still being used (I wish it were not - but its going to be a bit of a larger refactoring effort to move this to python).

jgfouca commented on driver_cpl/cime_config/testdefs/testlist_drv.xml:
```diff
@@ -672,4 +654,195 @@
       </machine>
     </machines>
   </test>
+=======
```
====== again

mvertens replied:
Thanks for catching this. The merge added the old schema to the new. Its now fixed.

jgfouca commented on scripts/Tools/compare_namelists.pl:
```diff
@@ -23,14 +23,18 @@
 my $cnt2=$#basenml;
 my $shiftbl=1;
 foreach $line (@nlfile){
```
I don't think this is used anymore.

mvertens replied:
Do you mean that compare_namelists.pl is not used anymore?

jgfouca replied:
Yes, compare_namelists.pl should not be used anymore and should be deleted.

jgfouca commented on utils/python/CIME/case_st_archive.py:
```diff
@@ -178,7 +178,8 @@ def get_histfiles_for_restarts(case, archive, archive_entry, restfile):
         cmd = "ncdump -v %s %s " %(rest_hist_varname, os.path.join(rundir, restfile))
         rc, out, error = run_cmd(cmd, ok_to_fail=True)
-        expect(rc == 0, "Command %s failed rc=%d\nout=%s\nerr=%s" %(cmd, rc, out, error))
+        if rc != 0:
```
It's only a warning if ncdump fails?

mvertens replied:
Yes. This variable is needed to tell you the variable name on the restart file that contains settings for what history files need to be in your exeroot on restart. What I did not catch earlier is that some components (like cam, clm) contain history restart files - and that variable will not be on those files. So at this point, I don't want to change the scheme of the the config_archive.xml  - so as a result we are just printing out a warning and continuing. I caught this in running the aux_clm45 tests with the latest cime.

rljacob commented:
What is this new driver_cpl/bld/tesdir directory for?

rljacob commented:
Is this BFB in the coupler?

jgfouca commented:
@mvertens I'm not qualified to review most of this very large PR, but I did my best. A couple changes needed then I'll merge.

ekluzek commented:
Hi Rob

Currently the driver build-namelist can not be exercised outside of the
context of setting up a case with the cime scripts. You can't just run it
standalone and see if it gives you a reasonable namelist. So I created a
test directory so that you could at least run the build-namelist on it's
own and look at the resulting namelist. This makes working with the driver
build-namelist faster and easier. It doesn't verify the namelist is
correct, nor does it check that the logic in build-namelist is working
correctly, but it does allow for a simple straightforward check.

*Erik Kluzek, (CGD at NCAR)*

*National Center for Atmospheric Research*

*Boulder CO, *

*(off) (303)497-1326 (fax) (303)497-1348*

*(skype) ekluzek    (cell) (303)859-0183*

*Pronouns: he/his/him*

*------------------ Home page ------------------------*

*          http://www.cgd.ucar.edu/~erik <http://www.cgd.ucar.edu/~erik>*

*          https://staff.ucar.edu/users/erik
<https://staff.ucar.edu/users/erik>*

*!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*

On Thu, Jun 30, 2016 at 4:52 PM, Robert Jacob <notifications@github.com>
wrote:

> What is this new driver_cpl/bld/tesdir directory for?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMCI/cime/pull/234#issuecomment-229811097>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ALAswFR0cQBkasC4er2YMt8xSmKWVEK4ks5qREipgaJpZM4JCp0n>
> .
>

mvertens commented:
Update the PR.

jgfouca commented:
@mvertens , I still the ====== in a couple files. Also, the perl compare_namelists should be deleted.

mvertens commented:
I only found ==== in testlist_allactive.xml. I removed the one in testlist_drv.xml and also removed compare_namelists.pl. I am not sure why you are still seeing those. I just cloned a clean copy of branch and they are not there.

jgfouca commented:
@mvertens the only update that I see to this branch since it was opened was a small commit by @quantheory (Fix machines error in detecting threaded builds.) Are you sure you're pushing your fixes to the right branch?

mvertens commented:
@jgfouca I think I fixed things.This was my fault - I thought I was pushing successfully - but it actually was giving me an error message which I did not bother to look at. Both Sean's and my changes to address your comments should be in the latest PR update.

mvertens commented:
@jgfouca Please hold off on merging this PR - I am find a problem in some of the aux cesm tests with this branch

mvertens commented:
@jgfouca I have implemented fixes and have successfully run the regression test. So please go ahead and resolve this PR if you are comfortable with everything.

mvertens commented:
I did just push my latest changes.

jgfouca commented:
Looks good, thanks.
